[ONNX] Add script/trace different flatten and move optional type tests to runtime

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83599
* #83598
* #83186
* #83201
* __->__ #83184

fix #78119

Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

1. Move Optional Type tests from no runtime to runtime, as it's supported by ONNXRUNTIME.
2. Add ignoreNone flag for output comparison of internal tests

titaiwangms committed:
add script/trace different flatten

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
This PR enables runtime test to consider None type from torch in script mode.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
This PR enables runtime test to consider None type from torch in script mode.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, and this PR enables runtime test to keep None type from torch in script mode.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, and this PR enables runtime test to keep None type from torch in script mode.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, and this PR enables runtime test to keep None type from torch in script mode.

[ghstack-poisoned]

BowenBao commented on torch/onnx/verification.py:
```diff
@@ -122,10 +122,14 @@ def _compare_ort_pytorch_outputs(
     atol: float,
     check_shape: bool,
     check_dtype: bool,
+    flatten: bool,
```
You'll probably need to rebase now that #82622 is in. Please consider appending to docstring explaining what `flatten` does. Because it is not very clear to me as there is also some sort of flattening being done when `flatten==False`.

AllenTiTaiWang replied:
`flatten` here actually stands for `tracing`(True) or `scripting`(False), and it it's tracing(True), we use torch.jit._flatten which ignores None type in torch output, but scripting(False) uses `_inline_flatten_list` which unpacks it and keep None type, and we need None type for Optional type tests.

It is a tricky part that `flatten` actually comes from https://github.com/pytorch/pytorch/blob/bb1e3d8008fc0e40133a3b08bcfb598cd084897b/test/onnx/onnx_test_common.py#L99-L117

I found `flatten` was decided by the model is traced or scripted, and this is exactly what I need, so I extend it all the way down to output processing. 

PLEASE ADVICE: This could be a really bad idea reusing a parameter that serves as other purpose, as it's mixing up information, but otherwise, I will have to add another parameter and pass it in from the very beginning.

shubhambhokare1 replied:
I think reusing the flatten parameter might be a bit confusing in this scenario, can we decide if a model is trace/script based on module type?

BowenBao replied:
Maybe name it `ignoreNone` or something similar?

AllenTiTaiWang replied:
Done.

AllenTiTaiWang replied:
Done.

AllenTiTaiWang replied:
I committed to the wrong PR, so I messed up this PR with `Move Optional Type to runtime` ... They are now squashed together... Please have another review on this Thanks!

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten for optional element"


Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

[ghstack-poisoned]

AllenTiTaiWang commented:
cc @BowenBao for final review

justinchuby commented on test/onnx/onnx_test_common.py:
```diff
@@ -96,7 +96,7 @@ def run_test(
         verbose=False,
     ):
-        def _run_test(m, remained_onnx_input_idx, flatten=True):
+        def _run_test(m, remained_onnx_input_idx, flatten=True, ignoreNone=True):
```
snake case

AllenTiTaiWang replied:
Done

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -132,6 +133,10 @@ def _compare_ort_pytorch_outputs(
         pt_outs: outputs from PyTorch.
         rtol (float, optional): relative tolerance in comparison between ONNX and PyTorch outputs.
         atol (float, optional): absolute tolerance in comparison between ONNX and PyTorch outputs.
+        ignoreNone (bool, optional): Default True. If True, ignore None type in
```
When writing docstring for booleans, use: "Whether to..."

> Whether to ignore None values in the torch output, ...

Mention the default in the end. Match the behavior in the signature since currently it is required.

AllenTiTaiWang replied:
Done

justinchuby reviewed (changes requested):
minor comments

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten and move optional type tests to runtime"


fix #78119

Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

1. Move Optional Type tests from no runtime to runtime, as it's supported by ONNXRUNTIME.
2. Add ignoreNone flag for output comparison of internal tests

[ghstack-poisoned]

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -122,6 +122,7 @@ def _compare_ort_pytorch_outputs(
     atol: float,
     check_shape: bool,
     check_dtype: bool,
+    ignore_none: bool,
```
If optional, I would annotate as Optional[bool]. Same is needed for rtol and atol.

AllenTiTaiWang replied:
Tested with lintrunner, and found these can't be optional as they are consumed later in the code.

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten and move optional type tests to runtime"


fix #78119

Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

1. Move Optional Type tests from no runtime to runtime, as it's supported by ONNXRUNTIME.
2. Add ignoreNone flag for output comparison of internal tests

[ghstack-poisoned]

AllenTiTaiWang commented on torch/onnx/verification.py:
```diff
@@ -158,7 +167,8 @@ def _compare_ort_pytorch_outputs(
                 # Allow different but broadcastable output shapes.
-                ort_out, pt_out = np.broadcast_arrays(ort_out, pt_out)
+                ort_out, pt_out_np = np.broadcast_arrays(ort_out, pt_out)
+                pt_out = torch.from_numpy(pt_out_np)
```
Had to add this, or lintrunner would yell at me, as we expect pt_out as Sequence[Tensor], but np.broadcast_arrays makes it numpy

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -579,17 +591,18 @@ def verify(
-    verbose: bool = False,
-    fixed_batch_size: bool = False,
-    use_external_data: bool = False,
+    verbose: Optional[bool] = False,
+    fixed_batch_size: Optional[bool] = False,
+    use_external_data: Optional[bool] = False,
```
These should not be Optional. Same for the `bool` args below

justinchuby commented on torch/onnx/verification.py:
```diff
@@ -309,6 +320,7 @@ def compare_ort_pytorch_model_with_input(input_args, input_kwargs):
             atol,
             check_shape,
             check_dtype,
+            ignore_none,
             accetable_error_persentage,
```
Use keyword args when calling named params:

`ignore_none=ignore_none` etc. Could you help fix these lines?

justinchuby replied:
Sorry my bad they are not named. Closed.

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten and move optional type tests to runtime"


fix #78119

Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

1. Move Optional Type tests from no runtime to runtime, as it's supported by ONNXRUNTIME.
2. Add ignoreNone flag for output comparison of internal tests

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten and move optional type tests to runtime"


fix #78119

Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

1. Move Optional Type tests from no runtime to runtime, as it's supported by ONNXRUNTIME.
2. Add ignoreNone flag for output comparison of internal tests

[ghstack-poisoned]

AllenTiTaiWang commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1599,9 +1600,9 @@ def forward(self, x: int, y: int):
-    # We don't know Optional.elem_type, so we can't construct a valid Optional.
-    # Tests for Optional outputs (control flow with None in one branch,
-    # not-None in another) are in test_pytorch_onnx_no_runtime.py.
+    # Issue 84130: ONNX ignores mustNone() node, while pytorch
```
track https://github.com/pytorch/pytorch/issues/84130.
This is a workaround here, but if any reviewer thinks this is a bad idea even for a temporary workaround, I understand.

AllenTiTaiWang replied:
cc @BowenBao What do you think? Can we merge this?

BowenBao replied:
sgtm

AllenTiTaiWang replied:
Thanks! I will work on this right away.

titaiwangms committed:
Update on "[ONNX] Add script/trace different flatten and move optional type tests to runtime"


fix #78119

Why:
As in onnx tests verification code, we used to only consider tracing output, which ignores None type, this PR enables runtime test to keep None type in torch in script mode.

1. Move Optional Type tests from no runtime to runtime, as it's supported by ONNXRUNTIME.
2. Add ignoreNone flag for output comparison of internal tests

[ghstack-poisoned]

AllenTiTaiWang commented:
@pytorchbot merge -g
